dace/libraries/blas/nodes/gemm.py

```diff
@@ -425,6 +425,575 @@ def expansion(node, state, sdfg):
         return tasklet
 
 
+@dace.library.expansion
+class ExpandGemmFPGA1DSystolic(ExpandTransformation):
+    """
+    FPGA based implementation of GEMM, using a 1D systolic array.
+
+    Currently it supports non-transposed input matrices, and non-vectorized input array A.
+    """
+
+    environments = []
+
+    @staticmethod
+    def expansion(node,
+                  parent_state,
+                  parent_sdfg,
+                  num_pes=32,
+                  tile_size_m=None):
+        '''
+        GEMM node expansion.
+
+        :param node: Node to expand.
+        :param parent_state: State that the node is in.
+        :param parent_sdfg: SDFG that the node is in.
+        :param num_pes: Number of Processing Elements of the systolic array. By default it is set to 32.
+
+        :param tile_size_m: tiling size considering columns of the input matrix B and resulting matrix C.
+                            If B/C are vectorized, the tile size refers to the vectorized container.
+                            If set to None, no tiling is used, corresponding to setting the tile size
+                            equal to the number of columns of B/C.
+        :return:
+        '''
+
+        ((edge_a, outer_array_a, shape_a, strides_a), (edge_b, outer_array_b,
+                                                       shape_b, strides_b),
+         cdata) = _get_matmul_operands(node, parent_state, parent_sdfg)
+
+        dtype_a = outer_array_a.dtype.type
+        dtype_b = outer_array_b.dtype.type
+        dtype_c = dace.DTYPE_TO_TYPECLASS[np.result_type(dtype_a, dtype_b).type]
+        shape_c = (shape_a[0], shape_b[1])
+        if node.transA:
+            raise NotImplementedError(
+                "GEMM FPGA expansion not implemented for transposed A.")
+        if node.transB:
+            raise NotImplementedError(
+                "GEMM FPGA expansion not implemented for transposed B.")
+
+        if outer_array_a.veclen > 1:
+            raise NotImplementedError(
+                "Vectorization not support for input array A.")
+
+        if len(shape_a) != 2 or len(shape_b) != 2 or shape_a[1] != shape_b[0]:
+            raise SyntaxError("Matrix sizes must match")
+
+        ######################################################################
+        # GEMM Parameters and checks
+
+        # Note: the following sizes consider also vectorization
+        vec_width = outer_array_b.dtype.veclen
+        vec_type = dace.vector(dtype_c, vec_width)
+        N, K, M = shape_a[0], shape_a[1], shape_b[1]
+
+        P = num_pes
+        T = tile_size_m
+        if T is None:
+            T = M
+
+        # we will perform sanity check using T and M. But at this stage, we still
+        # don't know to what outer symbol they will map.
+        # We try to resolve them to constant if they are symbolic, otherwise we skip the checks
+        def resolve_symbol_to_constant(symb: dace.symbolic.symbol, parent_sdfg):
```

Comment:
This seems useful, why not hoist it into a common file, like `sdfg.py` or `symbolic.py`?